src/master-playlist-controller.js

```diff
@@ -875,40 +875,55 @@ export class MasterPlaylistController extends videojs.EventTarget {
     const enabledPlaylists = playlists.filter(isEnabled);
     const isFinalRendition = enabledPlaylists.length === 1 && enabledPlaylists[0] === currentPlaylist;
 
-    if (playlists.length === 1) {
-      // Never blacklisting this playlist because it's the only playlist
+    // Don't blacklist the only playlist unless it was blacklisted
+    // forever
+    if (playlists.length === 1 && blacklistDuration !== Infinity) {
       videojs.log.warn('Problem encountered with the current ' +
                        'playlist. Trying again since it is the only playlist.');
 
       this.tech_.trigger('retryplaylist');
       return this.masterPlaylistLoader_.load(isFinalRendition);
     }
 
+    // Blacklist this playlist
+    currentPlaylist.excludeUntil = Date.now() + (blacklistDuration * 1000);
+    this.tech_.trigger('blacklistplaylist');
+    this.tech_.trigger({type: 'usage', name: 'hls-rendition-blacklisted'});
+
     if (isFinalRendition) {
       // Since we're on the final non-blacklisted playlist, and we're about to blacklist
       // it, instead of erring the player or retrying this playlist, clear out the current
       // blacklist. This allows other playlists to be attempted in case any have been
       // fixed.
-      videojs.log.warn('Removing all playlists from the blacklist because the last ' +
-                       'rendition is about to be blacklisted.');
+      let unblacklisted = false;
+
       playlists.forEach((playlist) => {
         if (playlist.excludeUntil !== Infinity) {
+          unblacklisted = true;
           delete playlist.excludeUntil;
         }
       });
-      // Technically we are retrying a playlist, in that we are simply retrying a previous
-      // playlist. This is needed for users relying on the retryplaylist event to catch a
-      // case where the player might be stuck and looping through "dead" playlists.
-      this.tech_.trigger('retryplaylist');
-    }
 
-    // Blacklist this playlist
-    currentPlaylist.excludeUntil = Date.now() + (blacklistDuration * 1000);
-    this.tech_.trigger('blacklistplaylist');
-    this.tech_.trigger({type: 'usage', name: 'hls-rendition-blacklisted'});
+      if (unblacklisted) {
+        videojs.log.warn('Removing all playlists from the blacklist because the last ' +
+                         'rendition is about to be blacklisted.');
+        // Technically we are retrying a playlist, in that we are simply retrying a previous
+        // playlist. This is needed for users relying on the retryplaylist event to catch a
+        // case where the player might be stuck and looping through "dead" playlists.
+        this.tech_.trigger('retryplaylist');
+      }
+    }
 
     // Select a new playlist
     const nextPlaylist = this.selectPlaylist();
+
+    if (!nextPlaylist) {
```

Comment:
It's possible for `selectPlaylist` to return nothing. We don't handle that case now because we make sure to always unblacklist before coming here. Now we don't do that because some playlists that cause playback to stall **should** stay blacklisted.